src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -1309,85 +1311,206 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
+                result = (differentBits == 0);
+                goto Result;
+            }
+#if TARGET_64BIT
+            else
+            {
+                nuint offset = length - sizeof(uint);
+                uint differentBits = LoadUInt(ref first) - LoadUInt(ref second);
+                differentBits |= LoadUInt(ref first, offset) - LoadUInt(ref second, offset);
+                result = (differentBits == 0);
+                goto Result;
+            }
+#endif
+        Longer:
+            // Only check that the ref is the same if buffers are large,
+            // and hence its worth avoiding doing unnecessary comparisons
+            if (!Unsafe.AreSame(ref first, ref second))
+            {
+                // C# compiler inverts this test, making the outer goto the conditional jmp.
+                goto Vector;
+            }
 
-                if (Vector.IsHardwareAccelerated && (byte*)lengthToExamine >= (byte*)Vector<byte>.Count)
+            // This becomes a conditional jmp foward to not favor it.
+            goto Equal;
+
+        Result:
+            return result;
+        // When the sequence is equal; which is the longest execution, we want it to determine that
+        // as fast as possible so we do not want the early outs to be "predicted not taken" branches.
+        Equal:
+            return true;
+
+        Vector:
+            if (Sse2.IsSupported)
+            {
+                if (Avx2.IsSupported && length >= (nuint)Vector256<byte>.Count)
                 {
-                    lengthToExamine -= Vector<byte>.Count;
-                    while ((byte*)lengthToExamine > (byte*)offset)
+                    Vector256<byte> vecResult;
+                    nuint offset = 0;
+                    nuint lengthToExamine = length - (nuint)Vector256<byte>.Count;
+                    if (lengthToExamine != 0)
                     {
-                        if (LoadVector(ref first, offset) != LoadVector(ref second, offset))
+                        do
                         {
-                            goto NotEqual;
-                        }
-                        offset += Vector<byte>.Count;
+                            vecResult = Avx2.CompareEqual(LoadVector256(ref first, offset), LoadVector256(ref second, offset));
+                            if (Avx2.MoveMask(vecResult) != -1)
+                            {
+                                goto NotEqual;
+                            }
+                            offset += (nuint)Vector256<byte>.Count;
+                        } while (lengthToExamine > offset);
                     }
-                    return LoadVector(ref first, lengthToExamine) == LoadVector(ref second, lengthToExamine);
-                }
 
-                Debug.Assert((byte*)lengthToExamine >= (byte*)sizeof(UIntPtr));
+                    // Do final compare as Vector256<byte>.Count from end rather than start
+                    Debug.Assert(lengthToExamine >= 0);
```

Comment:
Since _lengthToExamine_ is an unsigned data type this assertion will never trip so it can't detect errors in the logic.

Reply:
Done